src/main/java/com/hazelcast/aws/AwsClientConfigurator.java

```diff
@@ -0,0 +1,179 @@
+    private static final String ECS_SERVICE_NAME = "ecs";
+
+    private AwsClientConfigurator() {
+    }
+
+    static AwsClient createAwsClient(AwsConfig awsConfig) {
+        Environment environment = new Environment();
+        AwsMetadataApi metadataApi = new AwsMetadataApi(awsConfig);
+
+        String region = resolveRegion(awsConfig, metadataApi, environment);
+        validateRegion(region);
+
+        AwsCredentialsProvider credentialsProvider = new AwsCredentialsProvider(awsConfig, metadataApi, environment);
+        AwsEc2Api ec2Api = createEc2Api(awsConfig, region);
+
+        // EC2 Discovery
+        if (explicitlyEc2Configured(awsConfig) || (!explicitlyEcsConfigured(awsConfig) && !environment.isRunningOnEcs())) {
+            logEc2Environment(awsConfig, region);
+            return new AwsEc2Client(ec2Api, metadataApi, credentialsProvider);
+        }
+
+        // ECS Discovery
+        String cluster = resolveCluster(awsConfig, metadataApi, environment);
+        AwsEcsApi ecsApi = createEcsApi(awsConfig, region);
+        logEcsEnvironment(awsConfig, region, cluster);
+        return new AwsEcsClient(cluster, ecsApi, ec2Api, metadataApi, credentialsProvider);
+    }
+
+    static String resolveRegion(AwsConfig awsConfig, AwsMetadataApi metadataApi, Environment environment) {
+        if (isNotEmpty(awsConfig.getRegion())) {
+            return awsConfig.getRegion();
+        }
+
+        if (environment.isRunningOnEcs()) {
+            return environment.getAwsRegionOnEcs();
+        }
+
+        String availabilityZone = metadataApi.availabilityZoneEc2();
+        return availabilityZone.substring(0, availabilityZone.length() - 1);
+    }
+
+    private static AwsEc2Api createEc2Api(AwsConfig awsConfig, String region) {
+        String ec2Endpoint = resolveEc2Endpoint(awsConfig, region);
+        AwsRequestSigner ec2RequestSigner = new AwsRequestSigner(region, EC2_SERVICE_NAME);
+        return new AwsEc2Api(ec2Endpoint, awsConfig, ec2RequestSigner, Clock.systemUTC());
+    }
+
+    private static AwsEcsApi createEcsApi(AwsConfig awsConfig, String region) {
+        String ecsEndpoint = resolveEcsEndpoint(awsConfig, region);
+        AwsRequestSigner ecsRequestSigner = new AwsRequestSigner(region, ECS_SERVICE_NAME);
+        return new AwsEcsApi(ecsEndpoint, awsConfig, ecsRequestSigner, Clock.systemUTC());
+    }
+
+    static String resolveEc2Endpoint(AwsConfig awsConfig, String region) {
+        String ec2HostHeader = awsConfig.getHostHeader();
+        if (StringUtils.isEmpty(ec2HostHeader)
+            || ec2HostHeader.startsWith("ecs")
+            || ec2HostHeader.equals("ec2")
+        ) {
+            ec2HostHeader = DEFAULT_EC2_HOST_HEADER;
+        }
+        return ec2HostHeader.replace("ec2.", "ec2." + region + ".");
+    }
+
+    static String resolveEcsEndpoint(AwsConfig awsConfig, String region) {
+        String ecsHostHeader = awsConfig.getHostHeader();
```

Comment:
Do we even need `host-header` anymore? Maybe we could make it optional and enum-like with a choice of either `ecs` or `ec2`?

Reply:
As discussed on the stand up, I think we should keep it and consider removing (or changing to enum) in the next major release.